src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsBuilderExtensions.cs

```diff
@@ -58,22 +61,72 @@ public static IWebJobsBuilder UseHostId(this IWebJobsBuilder builder, string hos
 
         public static IWebJobsBuilder UseWebJobsStartup<T>(this IWebJobsBuilder builder) where T : IWebJobsStartup, new()
         {
-            return builder.UseWebJobsStartup(typeof(T));
+            return builder.UseWebJobsStartup<T>(NullLogger.Instance);
+        }
+
+        public static IWebJobsBuilder UseWebJobsStartup<T>(this IWebJobsBuilder builder, ILogger logger) where T : IWebJobsStartup, new()
+        {
+            return builder.UseWebJobsStartup(typeof(T), logger);
         }
 
         public static IWebJobsBuilder UseWebJobsStartup(this IWebJobsBuilder builder, Type startupType)
+        {
+            return builder.UseWebJobsStartup(startupType, NullLogger.Instance);
+        }
+
+        public static IWebJobsBuilder UseWebJobsStartup(this IWebJobsBuilder builder, Type startupType, ILogger logger)
         {
             if (!typeof(IWebJobsStartup).IsAssignableFrom(startupType))
             {
                 throw new ArgumentException($"The {nameof(startupType)} argument must be an implementation of {typeof(IWebJobsStartup).FullName}");
             }
 
             IWebJobsStartup startup = (IWebJobsStartup)Activator.CreateInstance(startupType);
-            startup.Configure(builder);
 
+            if (logger == NullLogger.Instance) 
+            {
+                startup.Configure(builder);
+            }
+            else
+            {
+                ConfigureAndLogUserConfiguredServices(startup, builder, logger);      
+            }
             return builder;
         }
 
+        private static void ConfigureAndLogUserConfiguredServices(IWebJobsStartup startup, IWebJobsBuilder builder, ILogger logger)
```

Comment:
Rather than receive an `ILogger`, I think we should take an `ILoggerFactory` (there's a `NullLoggerFactory.Instance` as well). That way, we control what the category of that logger is.

Then down here we can call `loggerFactory.CreateLogger<ServiceCollectionDecorator>` and we'll then be in control of exactly what the category is.

Reply:
That makes sense--I have one question though. If the logger is created within the scope of UseWebjobsStartup, how will I be able to access it in order to test that the expected services were logged? Is there an accessor to a factory's loggers?

Reply:
We have a `TestLoggerProvider` in the test code. So you create a `LoggerFactory`, add the test logger provider, then pass that `LoggerFactory` into your test. That will track Loggers created, all log messages, etc. If you search in the test code for where we use `TestLoggerProvider` you should find some examples.

Reply:
Ooh okay, I had tried that but registered a logger with the provider instead of using provider.GetAllLogMessages. Thank you!